### PR TITLE
refactor: Throwing error messages with field.

### PR DESCRIPTION
**Pull request by aimuz**
Path and field.Error will make the error message update accurately

Signed-off-by: aimuz <mr.imuz@gmail.com>

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, press enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind breaking
> /kind bug
> /kind cleanup
> /kind documentation
> /kind feature
> /kind hotfix

/kind breaking

**What this PR does / Why we need it**:

**Which issue(s) this PR fixes**:

Closes #

**Special notes for your reviewer**:

This PR uses field.ErrorList to record error messages. This will make the errors thrown more standard and easier to locate

### CODE REVIEW

**Comment by aimuz** [comment]
ref https://github.com/googleforgames/agones/pull/3103

Related to this, in order to describe more information in Message without having to perform type conversion

**Comment by aimuz** [comment]
https://console.cloud.google.com/cloud-build/builds/1735c2d3-1c58-44ca-8bd9-216e9edff77f;step=25?project=agones-images

Unable to see detailed logs

**Comment by markmandel** [comment]
> Unable to see detailed logs

Make sure you are using the correct google account when viewing the page.

**Comment by markmandel** [comment]
Thanks for all this work, but what's not clear to me is why this is an improvement?

What does this actually do?

**Comment by aimuz** [comment]
> What does this actually do?

There are several purposes for using the same validation as k8s. It allows for reuse of existing code and reduces development time.

The error messages thrown by this implementation are more detailed compared to the current one. For example, when creating a fleet with incorrect parameters, the current message is simply "Fleet configuration is invalid". This hides critical information that could be useful for debugging.

To address this issue, the proposed changes include implementing the `Error()` method in the `StatusError` struct of the `errors.go` file from the `k8s.io/apimachinery/pkg/api/errors` package. This method returns the error message stored in the `ErrStatus.Message` field.

By adopting this approach, it becomes easier to retrieve detailed error information through `err.Error()`, without requiring additional type assertions.

## Purpose of the PR

The motivation behind this pull request is to integrate Agones into a PaaS platform. However, during the integration process, there were errors occurring without clear descriptions, making it difficult to understand the root cause.

Currently, the program checks if the error is empty and then directly outputs the result of `Error()`. However, this method doesn't provide sufficient details about the error.

Through this PR, the aim is to make error handling more standardized and improve the clarity of error messages. The proposed changes in `errors.go` enable the creation of an `Invalid` error, which indicates that the item being processed is invalid. This error includes a list of causes, each containing information such as the cause type, message, and field.

For example, the introduction of this PR enables the following error message:

```json
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "Fleet.agones.dev \"AAsimple-game-server\" is invalid: metadata.name: Invalid value: \"AAsimple-game-server\": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')",
  "reason": "Invalid",
  "details": {
    "name": "AAsimple-game-server",
    "group": "agones.dev",
    "kind": "Fleet",
    "causes": [
      {
        "reason": "FieldValueInvalid",
        "message": "Invalid value: \"AAsimple-game-server\": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')",
        "field": "metadata.name"
      }
    ]
  },
  "code": 422
}
```

In this example, the error message clearly states that the `metadata.name` field has an invalid value according to the specified criteria.

Similarly, when the `allocationOverflow` field is not supported, the error message would be:

```json
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "admission webhook \"validations.agones.dev\" denied the request: Fleet configuration is invalid",
  "reason": "Invalid",
  "details": {
    "name": "simple-game-server",
    "group": "agones.dev",
    "kind": "Fleet",
    "causes": [
      {
        "reason": "FieldValueNotSupported",
        "message": "Value cannot be set unless feature flag FleetAllocationOverflow is enabled",
        "field": "allocationOverflow"
      }
    ]
  },
  "code": 400
}
```

As you can see, the error messages generated by kube-apiserver provide more detailed information. Although additional details are available in the `details` field, accessing them requires an extra layer of logic and type conversion.

The proposed changes aim to make errors more explicit and easier to understand.

**Comment by aimuz** [comment]
> Make sure you are using the correct google account when viewing the page.

https://console.cloud.google.com/cloud-build/builds/1735c2d3-1c58-44ca-8bd9-216e9edff77f;step=25?project=agones-images

This one can see the log

图片

This can't see the logs

https://console.cloud.google.com/cloud-build/builds/c26a61d2-b609-4bdb-b5d7-2d7ebba459b2?project=agones-images

图片

**Comment by aimuz** [comment]
@markmandel Do you have time to help me with the question of why I can't see the logs?

**Comment by markmandel** [comment]
Just to be 100% clear on the logs -- you can see the logs linked from this PR, but I'm guessing not the ones that are the e2e for each version of a gke cluster? (sub-builds)?

**Comment by markmandel** [comment]
> The error messages thrown by this implementation are more detailed compared to the current one. For example, when creating a fleet with incorrect parameters, the current message is simply "Fleet configuration is invalid". This hides critical information that could be useful for debugging.

Thank you very much for the detailed description -- this makes things much more clear, and this is sounding like a really good change.

> For example, the introduction of this PR enables the following error message:
> 
> ```json
> {
>   "kind": "Status",
>   "apiVersion": "v1",
>   "metadata": {},
>   "status": "Failure",
>   "message": "Fleet.agones.dev \"AAsimple-game-server\" is invalid: metadata.name: Invalid value: \"AAsimple-game-server\": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')",
>   "reason": "Invalid",
>   "details": {
>     "name": "AAsimple-game-server",
>     "group": "agones.dev",
>     "kind": "Fleet",
>     "causes": [
>       {
>         "reason": "FieldValueInvalid",
>         "message": "Invalid value: \"AAsimple-game-server\": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')",
>         "field": "metadata.name"
>       }
>     ]
>   },
>   "code": 422
> }
> ```
> 
> In this example, the error message clearly states that the `metadata.name` field has an invalid value according to the specified criteria.

Just because my memory is no good 😄 what sort of output did you see before this change?

Also, does this change improve things when kubectl fails?

Since this is a bigger change @zmerlynn - would you care to also pass your eyes over this, just for review?

File: pkg/apis/agones/v1/common.go
Comment (markmandel) [inline pkg/apis/agones/v1/common.go:95]: Looking at this now, what you propose makes lots of sense. Where before we would take the errors and convert them back into structs -- when we never had to do that.

Nice find.

File: pkg/apis/agones/v1/gameserver.go
Comment (markmandel) [inline pkg/apis/agones/v1/gameserver.go:557]: Curious - any reason this is a TODO without being implemented yet? Or were you waiting on some review before implementing this?

**Comment by markmandel** [comment]
Looking through the e2e tests (while you can't see them), here are some errors:

```
    gameserver_test.go:930: 
        	Error Trace:	/go/src/agones.dev/agones/test/e2e/gameserver_test.go:930
        	Error:      	Not equal: 
        	            	expected: "container"
        	            	actual  : "spec.template.spec.containers[0].resources.requests.memory"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-container
        	            	+spec.template.spec.containers[0].resources.requests.memory
        	Test:       	TestGameServerResourceValidation

  gameserver_test.go:939: 
        	Error Trace:	/go/src/agones.dev/agones/test/e2e/gameserver_test.go:939
        	Error:      	Not equal: 
        	            	expected: "container"
        	            	actual  : "spec.template.spec.containers[0].resources.requests.cpu"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-container
        	            	+spec.template.spec.containers[0].resources.requests.cpu
        	Test:       	TestGameServerResourceValidation

   fleet_test.go:1428: 
        	Error Trace:	/go/src/agones.dev/agones/test/e2e/fleet_test.go:1428
        	Error:      	Not equal: 
        	            	expected: "container"
        	            	actual  : "spec.template.spec.template.spec.containers[1].resources.requests.memory"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-container
        	            	+spec.template.spec.template.spec.containers[1].resources.requests.memory
        	Test:       	TestFleetResourceValidation

  fleet_test.go:1438: 
        	Error Trace:	/go/src/agones.dev/agones/test/e2e/fleet_test.go:1438
        	Error:      	Not equal: 
        	            	expected: "container"
        	            	actual  : "spec.template.spec.template.spec.containers[0].resources.requests.cpu"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-container
        	            	+spec.template.spec.template.spec.containers[0].resources.requests.cpu
        	Test:       	TestFleetResourceValidation
    fleet_test.go:1632: 
        	Error Trace:	/go/src/agones.dev/agones/test/e2e/fleet_test.go:1632
        	            				/go/src/agones.dev/agones/test/e2e/fleet_test.go:1440
        	Error:      	Should be true
        	Test:       	TestFleetResourceValidation
        	Messages:   	Was not able to find 'Request must be less than or equal to cpu limit'
    fleet_test.go:1632: 
        	Error Trace:	/go/src/agones.dev/agones/test/e2e/fleet_test.go:1632
        	            				/go/src/agones.dev/agones/test/e2e/fleet_test.go:1441
        	Error:      	Should be true
        	Test:       	TestFleetResourceValidation
        	Messages:   	Was not able to find 'Resource cpu limit value must be non negative'
    fleet_test.go:1632: 
        	Error Trace:	/go/src/agones.dev/agones/test/e2e/fleet_test.go:1632
        	            				/go/src/agones.dev/agones/test/e2e/fleet_test.go:1442
        	Error:      	Should be true
        	Test:       	TestFleetResourceValidation
        	Messages:   	Was not able to find 'Request must be less than or equal to memory limit'
```

That should keep you going 😄

**Comment by markmandel** [comment]
#3239 should fix the access issue for future builds once merged.

File: pkg/apis/agones/v1/gameserver.go
Comment (aimuz) [reply pkg/apis/agones/v1/gameserver.go:557]: APIHOOK is an interface, and I plan to modify the implementation of the interface in the next phase to avoid making too many changes in this PR.

If you think you can do it together in this PR, I will do the work together.

Comment (markmandel) [reply pkg/apis/agones/v1/gameserver.go:557]: Got it - that's fine - just wanted to check 👍🏻

**Comment by aimuz** [comment]
> That should keep you going 😄

Thanks, I will continue to complete this work

File: test/e2e/fleet_test.go
Comment (aimuz) [inline test/e2e/fleet_test.go:870]: This field is not yet available in metav1, so this direct type conversion

Comment (aimuz) [reply test/e2e/fleet_test.go:870]: https://github.com/kubernetes/kubernetes/issues/119152

I asked this question at kubernetes

File: pkg/apis/agones/v1/common.go
Comment (zmerlynn) [inline pkg/apis/agones/v1/common.go:94]: nit: here you declare var and then append to it, whereas on line 114 below you start with `allErrs := `. I actually prefer the line 114 version since it's one fewer line, but regardless I would rather they be consistent. :)

File: pkg/apis/agones/v1/gameserver.go
Comment (zmerlynn) [reply pkg/apis/agones/v1/gameserver.go:557]: Just a workflow nit - I don't see it much in this project, but consider using something like `TODO(#xyz)` where `xyz` is an issue number. Then you can search for it easily to ensure you've accounted for all the TODOs when you're working on a long arc of PRs.

**Review by zmerlynn — COMMENTED** [review]
This LGTM, great change! I left a couple of nits, but lets try to get it in soon since it's such a big refactor (I don't want you to have to rebase too much).

File: pkg/apis/agones/v1/common.go
Comment (aimuz) [reply pkg/apis/agones/v1/common.go]: This is my omission, it's done

File: pkg/apis/agones/v1/gameserver.go
Comment (aimuz) [inline pkg/apis/agones/v1/gameserver.go:557]: https://github.com/googleforgames/agones/pull/3239#discussion_r1255006253

**Comment by aimuz** [comment]
> Just because my memory is no good 😄 what sort of output did you see before this change?

@markmandel I'm sorry I ignored this message

This is the content before modification

```yaml
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "admission webhook \"validations.agones.dev\" denied the request: GameServer configuration is invalid",
  "reason": "Invalid",
  "details": {
    "name": "simple-game-server-ghnf7",
    "group": "agones.dev",
    "kind": "GameServer",
    "causes": [
      {
        "reason": "FieldValueInvalid",
        "message": "ContainerPort cannot be specified with Passthrough PortPolicy",
        "field": "default.containerPort"
      }
    ]
  },
  "code": 400
}
```

Changes to this PR

```json
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "admission webhook \"validations.agones.dev\" denied the request: GameServer.agones.dev \"simple-game-server-x8w4m\" is invalid: spec.ports[0].containerPort: Required value: ContainerPort cannot be specified with Passthrough PortPolicy",
  "reason": "Invalid",
  "details": {
    "name": "simple-game-server-x8w4m",
    "group": "agones.dev",
    "kind": "GameServer",
    "causes": [
      {
        "reason": "FieldValueRequired",
        "message": "Required value: ContainerPort cannot be specified with Passthrough PortPolicy",
        "field": "spec.ports[0].containerPort"
      }
    ]
  },
  "code": 422
}
```

**Review by zmerlynn — APPROVED** [review]

**Review by markmandel — APPROVED** [review]
Thanks for the error message diff - look great!

**Comment by markmandel** [comment]
Oops, got caught when we were recreating the e2e clusters. Let's hit retry when complete.

**Comment by aimuz** [comment]
Do I need to do something?

**Comment by markmandel** [comment]
Nah, we've got it.

**Comment by aimuz** [comment]
https://console.cloud.google.com/cloud-build/builds/bc4ae54d-10b7-4274-b9ee-f91dc15e282d?project=agones-images

It looks like the timeout is causing the error, do I need to resubmit to trigger the retest?

**Comment by gongmax** [comment]
kicked off another retry

**Pull request merged**
